src/parser.rs

```diff
@@ -104,4 +128,24 @@ pub(crate) mod tests {
         assert_eq!(serialize_for_test(&children), "<tbody></tbody>");
         Ok(())
     }
+
+    #[tokio::test]
+    async fn test_document_parse_errors() -> io::Result<()> {
+        let result =
+            parse_document_async("<!DOCTYPE html>Hello <strong><em>world</strong></em>".as_bytes())
+                .await;
+        assert!(matches!(result, Err(e) if e.kind() == io::ErrorKind::InvalidData));
```

Comment:
It might be good to test that the error message actually contains the line number...